### PR TITLE
Added back screenshot functionality

### PR DESCRIPTION
**Pull request by eneoli**
Added back the screenshot functionality by using grim and slurp under the hood.

includes all functions from `org.freedesktop.impl.portal.Screenshot` (Screenshot, PickColor)

Fixes: #124

### CODE REVIEW

File: src/portals/Screenshot.cpp [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+    call >> appID;
+
+    std::string parentWindow;
+    call >> parentWindow;
+
+    Debug::log(LOG, "[screenshot] New PickColor request:");
+    Debug::log(LOG, "[screenshot]  | {}", requestHandle.c_str());
+    Debug::log(LOG, "[screenshot]  | appid: {}", appID);
+
+    const std::string PICK_COLOR_CMD = "grim -g \"$(slurp -p)\" -t ppm -";
```
Comment (vaxerski) [inline src/portals/Screenshot.cpp:77]: why don't we use our own hyprpicker, if the user has it installed, and only fall back to slurp?

File: src/portals/Screenshot.cpp [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+    Debug::log(LOG, "[screenshot] New screenshot request:");
+    Debug::log(LOG, "[screenshot]  | {}", requestHandle.c_str());
+    Debug::log(LOG, "[screenshot]  | appid: {}", appID);
+
+    bool isInteractive = options.count("interactive") && options["interactive"].get<bool>();
+
+    // make screenshot
+    const std::string FILE_PATH = "/tmp/xdph_screenshot.png";
+    const std::string SNAP_CMD  = "grim " + FILE_PATH;
+    const std::string SNAP_INTERACTIVE_CMD = "grim -g \"$(slurp)\" " + FILE_PATH;
```
Comment (vaxerski) [inline src/portals/Screenshot.cpp:42]: I would, at launch, in init, check for these programs' existence. If they don't exist, don't create the portal.

File: src/portals/Screenshot.cpp [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+
+    std::unordered_map<std::string, sdbus::Variant> results;
+    results["uri"] = "file://" + FILE_PATH;
+
+    remove(FILE_PATH.c_str());
```
Comment (vaxerski) [inline src/portals/Screenshot.cpp:47]: `std::filesystem` please

File: src/helpers/MiscFunctions.cpp [hunk 4]
```diff
@@ -24,4 +24,8 @@ void addHyprlandNotification(const std::string& icon, float timeMs, const std::s
     Debug::log(LOG, "addHyprlandNotification: {}", CMD);
     if (fork() == 0)
         execl("/bin/sh", "/bin/sh", "-c", CMD.c_str(), nullptr);
+}
+
+bool fileExists(std::string path) {
+    return access(path.c_str(), F_OK) == 0;
```
Comment (vaxerski) [inline src/helpers/MiscFunctions.cpp:30]: `std::filesystem` please

File: src/portals/Screenshot.cpp [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+    for (uint8_t i = 0; i < 2; i++) {
+        uint64_t next = rgbColor.find(' ');
+
+        if (next == std::string::npos) {
+            Debug::log(ERR, "hyprpicker returned strange output: " + rgbColor);
+            sendEmptyDbusMethodReply(call, 1);
+            return;
+        }
+
+        colors[i] = std::stoi(rgbColor.substr(0, next));
```
Comment (vaxerski) [inline src/portals/Screenshot.cpp:29]: all `std::stox` calls mandatory `try{}catch()` block

File: src/portals/Screenshot.cpp [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+    if (!ppmColor.starts_with("P6 1 1 ")) {
+        Debug::log(ERR, "grim did not return a PPM Image for us.");
+        sendEmptyDbusMethodReply(call, 1);
+        return;
+    }
+
+    // convert it to a rgb value
+    std::string maxValString = ppmColor.substr(7, ppmColor.size());
+    maxValString             = maxValString.substr(0, maxValString.find(' '));
+    uint32_t maxVal          = std::stoi(maxValString);
```
Comment (vaxerski) [inline src/portals/Screenshot.cpp:62]: ditto

File: src/portals/Screenshot.cpp [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+    bool slurpInstalled      = inShellPath("slurp");
+
+    if (!slurpInstalled && !hyprPickerInstalled) {
+        Debug::log(ERR, "Neither slurp nor hyprpicker found. We can't pick colors.");
+        sendEmptyDbusMethodReply(call, 1);
+        return;
+    }
+
+    // use hyprpicker if installed, slurp as fallback
+    if (inShellPath("hyprpicker")) {
```
Comment (vaxerski) [inline src/portals/Screenshot.cpp:175]: you can reuse `hyprPickerInstalled`

File: src/helpers/MiscFunctions.cpp [hunk 8]
```diff
@@ -24,4 +29,41 @@ void addHyprlandNotification(const std::string& icon, float timeMs, const std::s
...
+    std::vector<std::string> paths;
+    uint32_t                 nextBegin = 0;
+    for (uint32_t i = 0; i < pathString.size(); i++) {
+        if (path[i] == ':') {
+            paths.push_back(pathString.substr(nextBegin, i - nextBegin));
+            nextBegin = i + 1;
+        }
+    }
+
+    if (nextBegin < pathString.size()) {
```
Comment (vaxerski) [inline src/helpers/MiscFunctions.cpp:58]: style nit: no {} around short ifs

File: src/helpers/MiscFunctions.cpp [hunk 9]
```diff
@@ -24,4 +29,41 @@ void addHyprlandNotification(const std::string& icon, float timeMs, const std::s
     Debug::log(LOG, "addHyprlandNotification: {}", CMD);
     if (fork() == 0)
         execl("/bin/sh", "/bin/sh", "-c", CMD.c_str(), nullptr);
-}
\ No newline at end of file
+}
+
+bool inShellPath(const std::string& exec) {
+
+    if (exec.starts_with("/") || exec.starts_with("./") || exec.starts_with("../")) {
```
Comment (vaxerski) [inline src/helpers/MiscFunctions.cpp:36]: ditto

File: src/helpers/MiscFunctions.cpp [hunk 10]
```diff
@@ -24,4 +29,41 @@ void addHyprlandNotification(const std::string& icon, float timeMs, const std::s
...
+        return std::filesystem::exists(exec);
+    }
+
+    // we are relative to our PATH
+    const char* path = std::getenv("PATH");
+
+    if (!path) {
```
Comment (vaxerski) [inline src/helpers/MiscFunctions.cpp:43]: ye, and others.

**Review by vaxerski — COMMENTED** [review]
rest looks alright. Sorry for a late review, idfk github's notifications arent fucking showing up I hate this platform

**Comment by eneoli** [comment]
That's it I guess. This should resolve #124, no?

**Comment by vaxerski** [comment]
mhm, add a `Fixes: #124` to your MR description and github will close it on merge

**Review by vaxerski — APPROVED** [review]
lgtm! thanks :)

**Comment by eneoli** [comment]
Thanks for building Hyprland ^^

**Pull request merged**